docs/tactics.md

```diff
@@ -1019,60 +1019,84 @@ Transforms the goal into its contrapositive.
 ### norm_cast
 
 This tactic normalizes casts inside expressions.
-It is basically a simp tactic with a specific set of lemmas to move casts
-upwards in the expression.
-Therefore it can be used more safely as a non-terminating tactic.
-It also has special handling of numerals.
+It can be tought as call to the simplifier with a specific set of lemmas to
+move casts upwards in the expression.
+It also has special handling of numerals and a simple heuristic to simplify
+casts "past" binary operators.
+Contrary to simp, it should be safe to use as a non-terminating tactic.
 
 For instance, given an assumption
 ```lean
 a b : ℤ
-h : ↑a + ↑b < (10 : ℚ)
+h : (a + b : ℚ) < 10
 ```
 
 writing `norm_cast at h` will turn `h` into
 ```lean
 h : a + b < 10
 ```
 
-You can also use `exact_mod_cast`, `apply_mod_cast`, `rw_mod_cast`
-or `assumption_mod_cast`.
-Writing `exact_mod_cast h` and `apply_mod_cast h` will normalize the goal and h before using `exact h` or `apply h`.
+`norm_cast` comes with wrapper tactics:
+-`exact_mod_cast`
+-`apply_mod_cast`
+-`rw_mod_cast`
+-`assumption_mod_cast`.
+
+Writing `exact_mod_cast h` and `apply_mod_cast h` will normalize the goal and `h`
+before using `exact h` or `apply h`.
 Writing `assumption_mod_cast` will normalize the goal and for every
 expression `h` in the context it will try to normalize `h` and use
 `exact h`.
 `rw_mod_cast` acts like the `rw` tactic but it applies `norm_cast` between steps.
 
-These tactics work with three attributes,
-`elim_cast`, `move_cast` and `squash_cast`.
+`norm_cast` works with three attributes,
+-`elim_cast`
+-`move_cast`
+-`squash_cast`
 
-`elim_cast` is for elimination lemmas of the shape
-`Π ..., P ↑a1 ... ↑an = P a1 ... an`, for instance:
+`elim_cast` is for elimination lemmas of the following shapes
+-`Π ..., P ↑a1 ... ↑an = P a1 ... an`
+-`Π ..., P ↑a1 ... ↑an ↔ P a1 ... an`
 
+examples:
 ```lean
-int.coe_nat_inj' : ∀ {m n : ℕ}, ↑m = ↑n ↔ m = n
+@[elim_cast] theorem coe_nat_inj' {m n : ℕ} : (↑m : ℤ) = ↑n ↔ m = n
+
+@[elim_cast] theorem coe_int_denom (n : ℤ) : (n : ℚ).denom = 1
 
-rat.coe_int_denom : ∀ (n : ℤ), ↑n.denom = 1
+@[elim_cast] theorem cast_id : ∀ n : ℚ, ↑n = n
 ```
 
-`move_cast` is for compositional lemmas of the shape
-`Π ..., ↑(P a1 ... an) = P ↑a1 ... ↑an`, for instance:
+`move_cast` is for compositional lemmas of the following shapes
+-`Π ..., ↑(P a1 ... an) = P ↑a1 ... ↑an`
+-`Π ..., ↑(P a1 ... an) ↔ P ↑a1 ... ↑an`
```

Comment:
Does this even make sense? It looks like `P a1 ... an` has to be a `Prop`, since it's part of an iff. But then what does `↑(P a1 ... an)` mean?

Reply:
Quite often `P` in the LHS is not the same as in the RHS. So, if you use `set.nonempty` as a `coe : set α → Prop`, then there could be a lemma
```lean
@[move_cast] lemma coe_union : ↑(s ∪ t) ↔ ↑s ∨↑t := sorry
```

Reply:
I think at the time I wrote this I was thinking about cast from bool to Prop, but I'm not even sure there is currently a lemma in the move_cast attribute that fits the pattern.
also I definitely should replace the P in the RHS to Q or P' to make it clearer that it's not the same operator (often it's the same operator, but with different implicit arguments, for instance its addition in both expressions but not over the same type)